// _dapp/index.js
const bel = require('bel')
const csjs = require('csjs-inject')

const web3 = require('_web3')

const winningProposals = require('./winningProposals.js')
const newProposals = require('./newProposals.js')
const ExplanationBox = require('./explanationBox.js')
const applicationForm = require('./applicationForm.js')

module.exports = dapp

function dapp (data) {
  const { wallet, winners, proposals, contract: { ballot } } = data

  var switchViewsButton = bel`<div class=${css.switchView} onclick=${()=>switchView()}>Create new proposal</div>`
  var oldWinnersGallery = winningProposals(winners)
  var text = 'Vote for proposal and help us reward the projects that benefit the community!'
  var explanationBox = ExplanationBox(text)
  var listOfProposals = newProposals(proposals, ballot, wallet, switchViewsButton)

  var newProposalForm = applicationForm(ballot, wallet)

  return bel`
    <div class=${css.indexMain}>
      ${switchViewsButton}
      ${oldWinnersGallery}
      ${explanationBox}
      ${listOfProposals}
    </div>`

  function switchView () {
    var allProposalsView = document.querySelector('[class^="proposalsMain"]')
    var addNewProposalView = document.querySelector('[class^="formMain"]')
    if (allProposalsView) {
      switchViewsButton.style.opacity = 1
      var newView = newProposalForm
      var parent = allProposalsView.parentNode
      parent.replaceChild(newView, allProposalsView )
      explanationBox.innerText = 'Create new proposal and compete for the crypto funds!'
      switchViewsButton.innerText = 'Vote for proposal'
    } else if (addNewProposalView) {
      switchViewsButton.style.opacity = 1
      var newView = listOfProposals
      var parent = addNewProposalView.parentNode
      parent.replaceChild(newView, addNewProposalView )
      explanationBox.innerText = text
      switchViewsButton.innerText = 'Create new proposal'
    }
  }
}

const fontAwesome = bel`<link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" rel='stylesheet' type='text/css'>`
const font = bel`<link href="https://fonts.googleapis.com/css?family=Yanone+Kaffeesatz" rel="stylesheet">`
document.head.appendChild(font)
document.head.appendChild(fontAwesome)

const css = csjs`
  body {
    font-family: 'Yanone Kaffeesatz', sans-serif;
    display: flex;
    justify-content: center;
    margin: 0;
    background-color: #39456b;
  }
  .indexMain {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 100%;
  }
  .switchView {
    display: flex;
    align-self: flex-end;
    margin: 3% 2% 0 0;
    color: #b61114;
    font-size: 20px;
    padding: 1%;
    text-transform: uppercase;
    background-color: #e2e1dc;
  }
  .switchView:hover {
    cursor: pointer;
    background-color: #fcfbec;
  }
`
